src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,237 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+
+#include <NovelRT/Experimental/Graphics/Vulkan/Graphics.Vulkan.h>
+#include <NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.h>
+
+namespace NovelRT::Experimental::Graphics::Vulkan
+{
+    VulkanGraphicsDevice* VulkanGraphicsBuffer::GetDeviceInternal() const noexcept
+    {
+        // TODO: This feels like it might be a perf issue later.
+        return dynamic_cast<VulkanGraphicsDevice*>(GraphicsDeviceObject::GetDevice().get());
```

Comment:
Um.... this feels yuck but I'll ask - are we taking the weak pointer, getting it, and then casting it back to a raw pointer? Or are we actually just ensuring it's a `VulkanGraphicsDevice*` here?

Reply:
Also, what happens if this fails by chance? Just a big ree from the runtime? Or should this not fail in the idiomatic case?

Reply:
The VulkanGraphicsBuffer expects a shared pointer to a VulkanGraphicsDevice. This is then propagated into a weak pointer to a base graphics device, so this should never fail unless someone did something WILD.